### PR TITLE
CORTX-32738: rgw: rgw support for provisioner Gconf migration to consul

### PR DESCRIPTION
**Pull request by sachitanands**
Problem:
Provisioner team is planning to migrate file based Gconf to consule based Gconf.
In order to do this, rgw component needs some changes to support this migration. 

Solution:
There are changes required in way we fetch the endpoints values from Gconf e.g. consul endpoint, service endpoint ports.
Approach we took is,  
   a) for any key we want to fetch, first  we need to fetch num_client/num_endpoints key from Gconf.
   b) based on this number, we need to iterate over each key and get desired value of key.

Signed-off-by: Sachitanand Shelake <sachitanand.shelake@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by sachitanands** [comment]
started custom build : https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/7146/

**Comment by sachitanands** [comment]
retest this please

**Comment by sachitanands** [comment]
https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/7186/

File: src/rgw/setup/rgw.py [hunk 1]
```diff
@@ -101,9 +101,11 @@ def config(conf: MappedConf):
         if not os.path.exists(config_file):
             raise SetupError(errno.EINVAL, f'"{config_file}" config file is not present.')
 
+        Log.info('Generating SSL certs')
```
Comment (sachitanands) [inline src/rgw/setup/rgw.py:104]: will remove these logger statements during final deployment.
As of now added for debugging purpose only.

File: src/rgw/setup/rgw.py [hunk 2]
```diff
@@ -322,15 +334,36 @@ def _check_consul_connection(conf: MappedConf):
...
+        return value_list
+
+    @staticmethod
+    def _fetch_consul_endpoint_url(conf: MappedConf, endpoint_type: str):
         """Fetch endpoint url based on endpoint type from cortx config."""
-        endpoints = Rgw._get_cortx_conf(conf, confstore_endpoint_key)
-        endpoints_values = list(filter(lambda x: urlparse(x).scheme == endpoint_type, endpoints))
-        if len(endpoints_values) == 0:
+        Log.info('Fetching consul endpoint url using new gconf method.')
+        consul_endpoints = Rgw._get_gconf_key_list(conf, const.CONSUL_NUM_ENDPOINT_KEY,
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:355]: should be under try catch block, as it raises exception @ line 343

File: src/rgw/setup/rgw.py [hunk 3]
```diff
@@ -322,15 +334,36 @@ def _check_consul_connection(conf: MappedConf):
             raise SetupError(errno.EINVAL, f"Consul server {host:port} not reachable")
 
     @staticmethod
-    def _fetch_endpoint_url(conf: MappedConf, confstore_endpoint_key: str, endpoint_type: str):
+    def _get_gconf_key_list(conf: MappedConf, gconf_num_key:str, actual_gconf_key:str):
+        """Get value list of specified gconf key."""
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:338]: Am not able to visualize the key value(list) pairs here, can you add examples in code itself

File: src/rgw/setup/rgw.py [hunk 4]
```diff
@@ -490,18 +523,22 @@ def _update_rgw_config_with_endpoints(conf: MappedConf, endpoints: dict, instanc
     def _get_service_port(conf: MappedConf, protocol: str):
         """Return rgw service port value."""
         port = None
-        endpoints = conf.get(const.SVC_ENDPOINT_KEY)
-        if endpoints:
-            svc_endpoints = list(filter(lambda x: urlparse(x).scheme == protocol, endpoints))
-            port = urlparse(svc_endpoints[0]).port
+        Log.info('Fetching service port using new Gconf method.')
+        svc_endpoints = Rgw._get_gconf_key_list(conf, const.SVC_ENDPOINT_NUM_KEY,
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:527]: same, need to catch exception

File: src/rgw/setup/rgw.py (hunk 2 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:366]: @cdeshmukh If we see any issue while fetching the key-values from Gconf then it should terminate the deployment with error message hence we are just raising SetupError at 343 line and push this SetupError to caller without handling it try-catch block.

File: src/rgw/setup/rgw.py (hunk 3 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:342]: done

File: src/rgw/setup/rgw.py [hunk 5]
```diff
@@ -327,15 +338,43 @@ def _check_consul_connection(conf: MappedConf):
             raise SetupError(errno.EINVAL, f"Consul server {host:port} not reachable")
 
     @staticmethod
-    def _fetch_endpoint_url(conf: MappedConf, confstore_endpoint_key: str, endpoint_type: str):
+    def _get_gconf_key_list(conf: MappedConf, gconf_num_key:str, actual_gconf_key:str):
+        """Get value list of specified gconf key."""
+        # e.g. for single key (key - cortx>common>external>consul>endpoints), it may have multiple values as,
+        # values: [tcp://cortx-consul-server:8301, http://cortx-consul-server:8301].
+        # To get this list of values,
+        #     a) first get number of endpoints for this key using another key (cortx>common>external>consul>num_endpoints),
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:346]: Am not sure why this num_endpoints is even introduced, given we can just iterate over configured number of endpoints? @tanujashinde0405 @nitin-seagate

File: src/rgw/setup/rgw.py (hunk 3 above)
Comment (cdeshmukh) [reply src/rgw/setup/rgw.py:342]: Good examples

File: src/rgw/setup/rgw.py (hunk 2 above)
Comment (cdeshmukh) [reply src/rgw/setup/rgw.py:366]: Ok, consul is must now for RGW from this PR, given GConf will reside there. If yes, i understand

**Review by cdeshmukh — APPROVED** [review]

**Comment by shailesh-vaidya** [comment]
retest this please

File: src/rgw/setup/rgw.py (hunk 2 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:366]: In above code context, we are using consul endpoint to save rgw lock variable.hence we are fetching the consul endpoint.
its not related to this Gconf migration to consul.

File: src/rgw/setup/rgw.py (hunk 5 above)
Comment (tanujashinde0405) [reply src/rgw/setup/rgw.py:346]: @cdeshmukh , num_ was added in gconf for all lists as requested by @ujjwalpl . Also As fas as i know, you cannot ietrate over multiple values in consul if you don't know how many values are there. Hence num_ keys helpes to iterate over lists .

File: src/rgw/setup/rgw.py (hunk 1 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py]: done

**Comment by sachitanands** [comment]
Tested the patch, 

a)  1 Node deployment with yaml based Gconf : [1Node-success.PNG](https://jts.seagate.com/secure/attachment/527105/527105_1Node-success.PNG) 

b) RGW setup logs with yaml based Gconf : [rgw_Setup_with_yaml_gconf.log](https://jts.seagate.com/secure/attachment/527108/527108_rgw_Setup_with_yaml_gconf.log) 

c) 1 Node deployment with consul based Gconf : [Consul_based_deployment.PNG](https://jts.seagate.com/secure/attachment/527106/527106_Consul_based_deployment.PNG) 
  (Note: here I have updated local docker images by replacing gconf path from yaml to consul so that only rgw will use consul based Gconf and rest of the component will use yaml based Gconf.) 
   Reference code  : https://github.com/Seagate/cortx-rgw-integration/blob/main/src/rgw/setup/rgw_setup#L42 

d) RGW setup logs with Consul based Gconf : [rgw_setup_with_consul_gconf.log](https://jts.seagate.com/secure/attachment/527107/527107_rgw_setup_with_consul_gconf.log)

**Review by shriya-deshmukh — APPROVED** [review]

File: src/rgw/setup/rgw.py (hunk 5 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:332]: yes correct, earlier we use to pass key to yaml based Gconf, which was returning all available values as list in one command but here in consul based Gconf it does not return list hence we need to iterate over each index and get the value using key & index. 
To get this max no of index, we need to fetch this **num_** (num_endpoints, num_services) key in Gconf.

Comment (cdeshmukh) [reply src/rgw/setup/rgw.py:332]: When i look at consul cli interface, listing keys should help here, that way we dont need to depend on the number of entries, so we can improve i believe @tanujashinde0405 

https://www.consul.io/commands/kv/get#listing-keys
image


Anyways, there is no pointing holding on to this PR for GConf improvements

**Review by tanujashinde0405 — APPROVED** [review]

**Pull request merged**